tests/unit/test_charm.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+
+# pylint: disable=protected-access
+"""Test for charm hooks."""
+
+import unittest
+import unittest.mock
+from unittest.mock import PropertyMock, patch
+
+import ops
+import ops.testing
+
+import charm_state
+from charm import JenkinsAgentCharm
+
+ACTIVE_STATUS_NAME = "active"
+BLOCKED_STATUS_NAME = "blocked"
+MAINTENANCE_STATUS_NAME = "maintenance"
+WAITING_STATUS_NAME = "waiting"
+ERROR_STATUS_NAME = "error"
+
+
+def raise_exception(exception: Exception):
+    """Raise exception function for monkeypatching.
+
+    Args:
+        exception: The exception to raise.
+
+    Raises:
+        exception: .
+    """
+    raise exception
+
+
+class TestCharm(unittest.TestCase):
```

Comment:
Out of curiosity, would there be a reason why we're using class based tests and not functional? I don't mind the stylistic differences :D

Reply:
I looked and this seems to be the way most machine charms do tests, so I took this approach to be consistent with other machine charms. But I'd love to look at an example of a machine charm that does functional based tests. Do you have any repo in mind that uses the functional approach?

Reply:
Heads up. pydantic is preferred over unittest, iirc

Reply:
Since we're re-writing the charm, do you think it is a good time to move away from `unittest` and to `pytest`?